src/python/pants/backend/jvm/targets/import_jars_mixin.py

```diff
@@ -4,100 +4,22 @@
-    assert not (kwargs is not None and payload is not None), 'may not provide both kwargs and payload'
-
-    field_pos = 0 if kwargs is not None else 1
-    target_representation = kwargs or payload.as_dict()
-
-    def gen_specs():
-      for fields_tuple in cls.imported_jar_library_spec_fields():
-        for item in target_representation.get(fields_tuple[field_pos], ()):
-          # For better error handling, this simply skips over non-strings, but we catch them
-          # with a WrongTargetType in JarLibrary.to_jar_dependencies.
-          if not isinstance(item, string_types):
-            raise JarLibrary.ExpectedAddressError(
-              'expected imports to contain string addresses, got {found_class} instead.'
-              .format(found_class=type(item).__name__)
-            )
-          yield item
-
-    return list(gen_specs())
-
-  @memoized_property
-  def imported_jars(self):
-    """:returns: the string specs of JarDependencies referenced by imported_jar_library_specs
-    :rtype: list of JarDependency
-    """
-    return JarLibrary.to_jar_dependencies(self.address,
-                                          self.imported_jar_library_specs(payload=self.payload),
-                                          self._build_graph)
+  expected_target_constraint = SubclassesOf(JarLibrary)
 
   @memoized_property
-  def imported_jar_libraries(self):
-    """:returns: target instances for specs referenced by imported_jar_library_specs.
-    :rtype: list of JarLibrary
-    """
-    libs = []
-    for spec in self.imported_jar_library_specs(payload=self.payload):
-      resolved_target = self._build_graph.get_target_from_spec(spec,
-                                                               relative_to=self.address.spec_path)
-      if not resolved_target:
-        raise self.UnresolvedImportError(
-          'Could not find JarLibrary target {spec} referenced from {relative_to}'
-          .format(spec=spec, relative_to=self.address.spec))
-      if not isinstance(resolved_target, JarLibrary):
-        raise self.ExpectedJarLibraryError(
-          'Expected JarLibrary got {target_type} for jar imports in {spec} referenced from '
-          '{relative_to}'
-          .format(target_type=type(resolved_target), spec=spec,
-                  relative_to=self.address.spec))
-      libs.append(resolved_target)
-    return libs
-
-  @classmethod
-  def compute_dependency_specs(cls, kwargs=None, payload=None):
-    """Tack imported_jar_library_specs onto the traversable_specs generator for this target."""
-    for spec in super(ImportJarsMixin, cls).compute_dependency_specs(kwargs, payload):
-      yield spec
-
-    imported_jar_library_specs = cls.imported_jar_library_specs(kwargs=kwargs, payload=payload)
-    for spec in imported_jar_library_specs:
-      yield spec
+  def all_imported_jar_deps(self):
+    # TODO: figure out if this OrderedSet is necessary.
```

Comment:
It's always a good idea to keep dependencies ordered: especially with protobuf. So yea, would keep.